src/vos/README.md

```diff
@@ -962,3 +964,67 @@ And once the different levels of the trees have been initialized, at no point wi
+
+For the client, as long as it receives the leader reply with modificaction done, it will regard it as succeed in spite of whether the modification (and related DTX) is really committed or not.
+It is the leader duty to commit the 'committable' DTXs asynchronously some time later, or when 'committable' DTXs count exceeds some threshold, or piggybacked via dispatched RPCs because of potential conflict with some subsequent modifications.
+
+When an application wants to read something from an object with multiple replicas, the client can send the RPC to any replica.
+On the server side, if related DTX (corresponding to the target record) has been committed (valid for every replica) or is committable (only valid for leader), then such record can be returned to the client.
+Otherwise, if related DTX is 'prepared', and if the current replica is not the leader, it will notify the client to retry the RPC with the leader replica.
+For leader replica, the 'prepared' DTX is regared as non-ready, then the old version corresponding to the latest committed or committable DTX will be returned to the client.
+
+The DTX model is built inside DAOS container.
+Each container maintains its own DTX table that is organized as two B+tree in SCM: one is for active DTXs, the other is for committed DTXs.
+The following diagram is for the modification of multiple replicated object under DTX model.
+
+
+<a id="8a"></a>
+<b>Modify multiple replicated object under DTX model</b>
+
+![../../doc/graph/Fig_066.png](../../doc/graph/Fig_066.png ">Modify multiple replicated object under DTX model")
+
+<a id="8b"></a>
+
+### DTX Leader Election
+
+In the DTX model, leader is a special replica that takes more work than non-leaders, including:
+
+1. All modification RPCs will be sent to the leader that will make some necessary sanity check, then dispatch the modification requests to other replicas.
+
+2. If some non-leader replica hits 'prepared' DTX, it will notify the client to retry related read-only RPC with leader replcia.
+Then the leader replica will handle the read-only RPCs that touch uncommitted DTXs.
+
+So genearally, the server on which the leader replica resides will have more load than non-leaders.
+To avoid or reduce the unbalaced load caused by DTX leader, we need multiple leaders on different servers for modifying different objects or different dkeys in the same object. Some rules for the leader election:
+
+R1: If multiple objects share the same redundancy group, then it is better to make the leaders for different objects to be on different servers.
+
+R2: If a multiple replicated object has multiple redundancy groups, then the leaders in different redundancy groups should be on different servers.
+
+R3: In each redundancy group, elect the replica which server has the lowest failure sequence as the leader to avoid leader switch as much as possible.
```

Comment:
Servers that fail frequently should be avoided in leader selection to avoid frequent leader migration.